packages/flutter/lib/src/widgets/orientation_builder.dart

```diff
@@ -51,3 +51,57 @@ class OrientationBuilder extends StatelessWidget {
     return LayoutBuilder(builder: _buildWithConstraints);
   }
 }
+
+/// Builds a widget tree that can depend on the device's orientation.
+///
+/// The orientation is obtained from [MediaQuery.orientationOf], which reflects
+/// the actual device orientation as reported by the platform. This ensures
+/// consistency with [MediaQueryData.orientation] and correct behavior on
+/// foldable devices and other scenarios where the device orientation may differ
+/// from layout dimensions.
+///
+/// This is different from [OrientationBuilder], which determines orientation
+/// based on the parent widget's layout constraints (width vs height), not the
+/// device's physical orientation.
+///
+/// {@tool snippet}
+/// This example shows how to use [DeviceOrientationBuilder] to display
+/// different widgets based on the device orientation.
+///
+/// ```dart
+/// DeviceOrientationBuilder(
+///   builder: (BuildContext context, Orientation orientation) {
+///     return Text(orientation == Orientation.portrait
+///         ? 'Device is in Portrait'
+///         : 'Device is in Landscape');
+///   },
+/// )
+/// ```
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [OrientationBuilder], which builds based on parent widget's layout
+///    constraints rather than device orientation.
+///  * [MediaQueryData.orientation], which provides the device orientation
+///    directly from [MediaQuery].
+///  * [LayoutBuilder], which exposes the complete layout constraints.
+///
+/// This widget requires a [MediaQuery] ancestor to obtain the orientation.
+/// Typically, this is provided by [MaterialApp] or [WidgetsApp].
+class DeviceOrientationBuilder extends StatelessWidget {
```

Comment:
I was thinking that maybe we should move this to its own file, but I think I changed my mind and it's ok here. Given how small this file is, and how well the tests fit together below, this seems pretty good.